### PR TITLE
1860 trains, routes and route selection

### PR DESCRIPTION
**Pull request by roseundy**
This is going to be one of the biggest set of changes for 1860.
This implements:
- 1860 m+n trains
- All of the various and sundry (and non-standard) route rules
- Allowing one train to run through a tokened-out city
- Automatically finding best towns to run to
- UI change to allow selecting between halt subsidies and town revenues
- Support for save/restore of connections to tiles with multiple intra-tile paths

One of the biggest problems I faced was not only allowing the user to choose between halts and towns for each train, but preserving that choice for the next OR. This required modifying the run_routes action as well as operating_info.

New UI for selecting number of halts in route_selector:
route_selector

The following common files had to be changed:
- Moved game-specific logic from Engine::Route to Game::Base
- Engine::Connection::id to support "ambiguous" connections by appending node ids to hex name
- Engine::Tile and Engine::Part::Node to support linear naming of node parts (node_id)
- Engine::Route, Action::RunRoutes, View::RouteSelector => connection_hexes remain strings until use
- Engine::Route, Action::RunRoutes support for halts
- View::RouteSelector suport for halts
- Engine::OperatingInfo support for halts

### CODE REVIEW

**Comment by roseundy** [comment]
This is what the new JSON for run_routes can look like for a route with halts and multiple intra-tile paths:
json

File: assets/app/view/game/route_selector.rb [hunk 1]
```diff
@@ -149,16 +158,29 @@ def render
             ]),
             h(:tbody, trains),
           ]),
-          actions,
+          actions(render_halts),
         ])
       end
 
+      def halt_actions(route, revenue, subsidy)
```
Comment (tobymao) [inline assets/app/view/game/route_selector.rb:165]: would it be easier to just pay out max halts instead of having this action?

**Comment by tobymao** [comment]
```One of the biggest problems I faced was not only allowing the user to choose between halts and towns for each train, but preserving that choice for the next OR. This required modifying the run_routes action as well as operating_info.```

maybe we should just simplify it by not alloiwg it

File: lib/engine/connection.rb [hunk 2]
```diff
@@ -81,17 +81,23 @@ def matches?(connection_hexes)
...
-          junction_map = {}
-
-          # skip over paths that have a junction we've already seen
-          @paths.each do |path|
-            uniq_paths << path if !junction_map[path.a] && !junction_map[path.b]
-            junction_map[path.a] = true if path.a.junction?
-            junction_map[path.b] = true if path.b.junction?
-          end
+          # deal with ambiguous intra-tile path
+          if @paths&.one? && @paths[0].tile.ambiguous_connection?
```
Comment (tobymao) [inline lib/engine/connection.rb:85]: why would paths be nil?

File: lib/engine/connection.rb [hunk 3]
```diff
@@ -81,17 +81,23 @@ def matches?(connection_hexes)
...
+            node_ids = @paths[0].nodes.map(&:node_id).sort
```
Comment (tobymao) [inline lib/engine/connection.rb:86]: youc ould do 

first, last @paths[0].nodes.map9&:nodeid).sort

File: lib/engine/game/base.rb [hunk 4]
```diff
@@ -775,6 +779,105 @@ def check_overlap(routes)
         end
       end
 
+      def check_connected(route, token)
```
Comment (tobymao) [inline lib/engine/game/base.rb:782]: did you change anything here that i need to look at or is it just movement?

File: lib/engine/tile.rb [hunk 5]
```diff
@@ -239,6 +241,12 @@ def terrain
       @upgrades.flat_map(&:terrains).uniq
     end
 
+    # if tile has more than one intra-tile paths, those paths
+    # cannot be identified with just a hex name
+    def ambiguous_connection?
+      @ambiguous_connection ||= @paths.select { |p| p.nodes.size > 1 }.size > 1
```
Comment (tobymao) [inline lib/engine/tile.rb:247]: could you do @paths.count { } ?

File: assets/app/view/game/route_selector.rb (hunk 1 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb:165]: I feel it's a pretty important tactical decision about running for revenue or to help the company. The rules specifically call this out as a choice the director makes.

File: lib/engine/connection.rb (hunk 2 above)
Comment (roseundy) [reply lib/engine/connection.rb:85]: In initialize: @path = nil

File: lib/engine/connection.rb (hunk 3 above)
Comment (roseundy) [reply lib/engine/connection.rb:86]: OK

File: lib/engine/game/base.rb (hunk 4 above)
Comment (roseundy) [reply lib/engine/game/base.rb:782]: Just moving game logic out of route

File: lib/engine/tile.rb (hunk 5 above)
Comment (roseundy) [reply lib/engine/tile.rb:247]: Sure. Always learning something new...

File: lib/engine/connection.rb (hunk 2 above)
Comment (tobymao) [reply lib/engine/connection.rb:85]: then how come the line below doesn't fail?

Comment (roseundy) [reply lib/engine/connection.rb:85]: Good point.

File: assets/app/view/game/route_selector.rb (hunk 1 above)
Comment (tobymao) [reply assets/app/view/game/route_selector.rb:165]: hm i thought the choice was whether or not to use the halt or not. is there another choice? i don't think halts pay out, they only ever help the company. the choice would be to use the halt or not, you'd only not want to use it in very rare situations

Comment (roseundy) [reply assets/app/view/game/route_selector.rb:165]: In a real life game with a M+N train, you decide how the +N stops are split between towns and halts. Since you can choose between upgrading to another halt or to a town, presumably you'd want the choice to get the revenue from the town if they were available.

Comment (roseundy) [reply assets/app/view/game/route_selector.rb:165]: How strongly do you dislike how I have it now?

Comment (tobymao) [reply assets/app/view/game/route_selector.rb:165]: i'll look at it some more this weekend

File: assets/app/view/game/route_selector.rb [hunk 6]
```diff
@@ -56,6 +55,7 @@ def render
           store(:selected_route, route, skip: true)
         end
 
+        render_halts = false
```
Comment (tobymao) [inline assets/app/view/game/route_selector.rb:58]: render_halts = route.num_halts

File: assets/app/view/game/route_selector.rb [hunk 7]
```diff
@@ -95,7 +95,13 @@ def render
             td_props = { style: { paddingRight: '0.8rem' } }
 
             children << h('td.right', td_props, route.distance)
-            children << h('td.right', td_props, revenue)
+            if route.num_halts
+              render_halts = true
```
Comment (tobymao) [inline assets/app/view/game/route_selector.rb:99]: remove this

File: assets/app/view/game/route_selector.rb [hunk 8]
```diff
@@ -149,16 +158,29 @@ def render
...
 
+      def halt_actions(route, revenue, subsidy)
+        change_halts = lambda do
+          route.cycle_halts
+          store(:selected_route, route)
+        end
+
+        [
+          revenue,
+          h(:br),
```
Comment (tobymao) [inline assets/app/view/game/route_selector.rb:173]: i don't think we need a br, just use a style

File: lib/engine/config/game/g_1860.rb [hunk 9]
```diff
@@ -576,35 +576,35 @@ module G1860
   "trains": [
     {
       "name": "2+1",
-      "distance": 2,
+      "distance": [2, 1],
```
Comment (tobymao) [inline lib/engine/config/game/g_1860.rb:579]: why is distance an array?

look at the format of distance in g_1846

Comment (tobymao) [reply lib/engine/config/game/g_1860.rb:579]: how can your code work with distance in this format?

File: lib/engine/game/g_1860.rb [hunk 10]
```diff
@@ -239,7 +241,7 @@ def close_other_companies!(company)
       end
 
       def biggest_train(corporation)
-        corporation.trains.map(&:distance).max || 0
+        corporation.trains.map { |t| t.distance.first }.max || 0
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:244]: there is no such thing as distance first, the format doesn't look correct

File: lib/engine/route.rb [hunk 11]
```diff
@@ -4,9 +4,10 @@
 
 module Engine
   class Route
+    attr_accessor :num_halts
```
Comment (tobymao) [inline lib/engine/route.rb:7]: why is this an accesor?

File: lib/engine/route.rb [hunk 12]
```diff
@@ -4,9 +4,10 @@
...
     attr_reader :last_node, :phase, :train, :routes
 
-    def initialize(game, phase, train, connection_hexes: [], routes: [], override: nil)
+    def initialize(game, phase, train, connection_hexes: [], routes: [], override: nil, num_halts: nil)
```
Comment (tobymao) [inline lib/engine/route.rb:10]: should we just call this halts?

File: lib/engine/game/g_1860.rb [hunk 13]
```diff
@@ -356,6 +359,175 @@ def legal_tile_rotation?(_entity, _hex, tile)
...
+        n_towns = route.stops.select { |n| n.town? && !n.halt? }.size
+        "#{n_cities}+#{n_towns}"
+      end
+
+      def revenue_for(route, stops)
+        stops.sum { |stop| stop.route_base_revenue(route.phase, route.train) }
+      end
+
+      def subsidy_for(_route, stops)
+        stops.select(&:halt?).size * HALT_SUBSIDY
```
Comment (tobymao) [inline lib/engine/game/g_1860.rb:521]: stops.count(&:halt?)

Comment (tobymao) [reply lib/engine/game/g_1860.rb:521]: but how does this work with num_halts in route?

File: lib/engine/route.rb [hunk 14]
```diff
@@ -270,8 +255,8 @@ def find_connections(connections_a, connections_b, other_paths)
     private
 
     def restore_connections(connection_hexes)
-      possibilities = connection_hexes.map do |hexes|
-        hex_ids = hexes.map(&:id)
+      possibilities = connection_hexes.map do |hex_ids|
+        hexes = hex_ids.map { |hex_str| @game.hex_by_id(hex_str.split.first) }
```
Comment (tobymao) [inline lib/engine/route.rb:259]: why did you change this to get the hexes here every time instead of on init?

File: lib/engine/tile.rb [hunk 15]
```diff
@@ -90,7 +90,8 @@ def self.part(type, params, cache)
                               visit_cost: params['visit_cost'],
                               route: params['route'],
                               format: params['format'],
-                              loc: params['loc'])
+                              loc: params['loc'],
+                              node_id: cache.size)
```
Comment (tobymao) [inline lib/engine/tile.rb:94]: why do you need to add a node_id?

File: assets/app/view/game/route_selector.rb (hunk 6 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb:58]: Ok

Comment (roseundy) [reply assets/app/view/game/route_selector.rb:58]: Ok

File: assets/app/view/game/route_selector.rb (hunk 7 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb:99]: Ok

File: assets/app/view/game/route_selector.rb (hunk 8 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb:173]: Ok, I'll have to figure out how I can use a style to place the button under the revenue.

File: lib/engine/config/game/g_1860.rb (hunk 9 above)
Comment (roseundy) [reply lib/engine/config/game/g_1860.rb:579]: M+N train => distance = [M, N]
Since I was rewriting the code dealing with distances, I chose to use a simple array over the more complicated hash used by 1846 and the others. In addition, the existing code and the hash structure couldn't handle how 1860 train distances work - specifically how an unused city/offboard allowance can be used for towns/halts (i.e. a run than doesn't use all M cities).

File: lib/engine/game/g_1860.rb (hunk 10 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:244]: distance is an array here.

File: lib/engine/route.rb (hunk 11 above)
Comment (roseundy) [reply lib/engine/route.rb:7]: Because on a brand new route that contains halts, num_halts is calculated and set by @game.compute_stops. Effectively the default num_halts for a new route is set to a number that maximizes revenue first, subsidies second.

File: lib/engine/route.rb (hunk 12 above)
Comment (roseundy) [reply lib/engine/route.rb:10]: Yeah, probably.

File: lib/engine/game/g_1860.rb (hunk 13 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:521]: route.num_halts is used by @game.compute_halts to determine how many stops are at halts.

File: lib/engine/route.rb (hunk 14 above)
Comment (roseundy) [reply lib/engine/route.rb:259]: restore_connection is only called on init, so I'm not sure what you mean by "every time"
I moved the @game.hex_by_id lookup here from Action::run_routes. It shouldn't be called any more often.

I moved it here because I needed to leave connections_hexes in string format so that hex_ids was the same string value stored in the JSON so that the Connection::id generated for "ambiguous" intra-tile paths would match correctly.

File: lib/engine/tile.rb (hunk 15 above)
Comment (roseundy) [reply lib/engine/tile.rb:94]: Because I needed an id that was unique across all forms of nodes for use in disambiguating nodes in intra-tile paths/connections. Tile generates indexes per class, but those indexes aren't unique across the node classes (i.e. there can be a City with index 0 as well as a Town with index 0).

File: lib/engine/route.rb (hunk 11 above)
Comment (roseundy) [reply lib/engine/route.rb:7]: This is how num_halts ever becomes anything other than nil for a new route.

File: lib/engine/tile.rb (hunk 15 above)
Comment (roseundy) [reply lib/engine/tile.rb:94]: I also needed node_id to be completely deterministic and based on the tile definition so that upgraded tiles will work correctly with existing routes/connections.

File: assets/app/view/game/route_selector.rb (hunk 6 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb:58]: Wait, render_halts is calculated across *all* routes. If any of the routes has halts, then render_halts is set true.

File: lib/engine/route.rb (hunk 12 above)
Comment (roseundy) [reply lib/engine/route.rb]: I renamed num_halts to halt

File: assets/app/view/game/route_selector.rb (hunk 8 above)
Comment (roseundy) [reply assets/app/view/game/route_selector.rb]: Changed to use div, which avoids the need for br

File: lib/engine/config/game/g_1860.rb (hunk 9 above)
Comment (tobymao) [reply lib/engine/config/game/g_1860.rb:579]: i don't think that's true. the current hash used in 1846 should allow for city and offboard overflowing to towns and halts.

File: lib/engine/game/g_1860.rb (hunk 13 above)
Comment (tobymao) [reply lib/engine/game/g_1860.rb:521]: doesn't that mean that this method should check num_halts and not select halts? don't you need to know which ones the user chose?

File: lib/engine/tile.rb (hunk 15 above)
Comment (tobymao) [reply lib/engine/tile.rb:94]: let's get rid of node_id and just change the part.index = index to not group by class, would that work?

File: lib/engine/config/game/g_1860.rb (hunk 9 above)
Comment (roseundy) [reply lib/engine/config/game/g_1860.rb:579]: OK, I'll look at the code again. I wasn't seeing how that could work.

File: lib/engine/game/g_1860.rb (hunk 13 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb:521]: All halts are the same, I just need to know how many halts are being stopped at. I could use either Route::halts or count the stops that are halts. They should always be the same.

File: lib/engine/tile.rb (hunk 15 above)
Comment (roseundy) [reply lib/engine/tile.rb:94]: Will the be guaranteed to result in the same index for a node after an upgrade? Since node_id is a function of the tile definition, as long as the node appears in the same order in both the original tile definition and it's upgraded tile, it will work correctly.

Comment (roseundy) [reply lib/engine/tile.rb:94]: It will work as long as all city and town definitions occur before any paths are defined. I can live with that.

Comment (roseundy) [reply lib/engine/tile.rb]: Done.

File: lib/engine/config/game/g_1860.rb (hunk 9 above)
Comment (roseundy) [reply lib/engine/config/game/g_1860.rb]: Done.

File: lib/engine/game/g_1860.rb (hunk 10 above)
Comment (roseundy) [reply lib/engine/game/g_1860.rb]: Moved to standard train definitions.

File: lib/engine/connection.rb [hunk 16]
```diff
@@ -81,17 +81,23 @@ def matches?(connection_hexes)
...
-
-          # skip over paths that have a junction we've already seen
-          @paths.each do |path|
-            uniq_paths << path if !junction_map[path.a] && !junction_map[path.b]
-            junction_map[path.a] = true if path.a.junction?
-            junction_map[path.b] = true if path.b.junction?
-          end
+          # deal with ambiguous intra-tile path
+          if @paths.one? && @paths[0].tile.ambiguous_connection?
+            node0, node1 = @paths[0].nodes.map(&:index).sort
```
Comment (tobymao) [inline lib/engine/connection.rb:86]: should this be id and not index?

Comment (roseundy) [reply lib/engine/connection.rb:86]: No, because id includes the tile number which would break the JSON stored route on upgrades.

**Review by tobymao — APPROVED** [review]
